torch/nn/functional.py

```diff
@@ -1827,8 +1827,22 @@ def multilabel_soft_margin_loss(input, target, weight=None, size_average=None,
     """
     if size_average is not None or reduce is not None:
         reduction = _Reduction.legacy_get_string(size_average, reduce)
-    input = torch.sigmoid(input)
-    return binary_cross_entropy(input, target, weight, None, None, reduction)
+
+    loss = -(target * logsigmoid(input) + (1 - target) * logsigmoid(-input))
+
+    if weight is not None:
+        loss = loss * weight
+
+    loss = loss.sum(dim=1)  # only return N loss values
+
+    if reduction == 'none':
+        return loss
```

Comment:
I'm still a bit confused by the behavior we have here. It looks like if reduction is none, we're returning a tensor of N losses, and each individual loss is the sum of the losses for each class of that sample. However, when reduction is elementwise_mean or sum, it seems that you're now averaging the losses for each class, rather than summing them. We should choose one of these and stick to it, I believe summing is the legacy behavior.

Reply:
oh, you're right. Yeah, I will change that to follow the legacy behavior, so that 
```
loss = F.multilabel_soft_margin_loss(outputs, labels, reduction='none')
loss_mean = F.multilabel_soft_margin_loss(outputs, labels, reduction='elementwise_mean')
loss_sum = F.multilabel_soft_margin_loss(outputs, labels, reduction='sum')

loss.sum() == loss_sum  # True
loss.mean() == loss_mean  # True
```

Reply:
Unless a loss is storing its elements in a weird way, the general rule of thumb is: if `loss` is the unreduced loss, then `loss.mean()` is the elementwise and `loss.sum()` is the sum.